### PR TITLE
[SMALLFIX][branch-1.6] Fix a race condition on netty block server

### PR DESCRIPTION
**Pull request by apc999**

### CODE REVIEW

File: core/server/worker/src/test/java/alluxio/worker/netty/UfsFileWriteHandlerTest.java [hunk 1]
```diff
@@ -50,7 +50,8 @@ public void before() throws Exception {
     UfsInfo ufsInfo = new UfsInfo(Suppliers.ofInstance(mockUfs), AlluxioURI.EMPTY_URI);
     Mockito.when(ufsManager.get(TEST_MOUNT_ID)).thenReturn(ufsInfo);
     Mockito.when(mockUfs.create(Mockito.anyString(), Mockito.any(CreateOptions.class)))
-        .thenReturn(mOutputStream);
+        .thenReturn(mOutputStream)
+        .thenReturn(new FileOutputStream(mFile, true));
```
Comment (calvinjia) [inline core/server/worker/src/test/java/alluxio/worker/netty/UfsFileWriteHandlerTest.java]: This seems a bit sketchy, should we just update this (or mock separately) in the individual test which needs it?

Comment (apc999) [reply core/server/worker/src/test/java/alluxio/worker/netty/UfsFileWriteHandlerTest.java]: I tried a few but don't see an obvious better way to mock this separately just for a test.

**Comment by apc999** [comment]
@calvinjia @yupeng9  PTAL

File: core/server/worker/src/main/java/alluxio/worker/netty/AbstractWriteHandler.java [hunk 2]
```diff
@@ -61,6 +61,9 @@
  *    an exception. When one of these 3 happens, a special packet is pushed to the buffer.
  * 2. The packet writer thread keeps polling packets from the buffer and processes them.
  *    NOTE: it is guaranteed that there is only one packet writer thread active at a given time.
+ * 3. Once a complete or cancel response is sent, there is no guarantee if the context instance
+ *    is represents the next arriving packet. Therefore, no further modification after response
```
Comment (yupeng9) [inline core/server/worker/src/main/java/alluxio/worker/netty/AbstractWriteHandler.java]: is present?

Comment (yupeng9) [reply core/server/worker/src/main/java/alluxio/worker/netty/AbstractWriteHandler.java]: Also grammar error on the last sentence

File: core/server/worker/src/main/java/alluxio/worker/netty/AbstractWriteHandler.java [hunk 3]
```diff
@@ -121,21 +124,24 @@ public void channelRead(ChannelHandlerContext ctx, Object object) throws Excepti
     Protocol.WriteRequest writeRequest = msg.getMessage().asWriteRequest();
 
     try (LockResource lr = new LockResource(mLock)) {
-      boolean isPreviousContextNull = mContext == null;
-      if (isPreviousContextNull) {
-        // When mContext is null, create a new one as catching exceptions and replying errors
+      boolean isNewContextCreated = false;
+      if (mContext == null || mContext.isDoneUnsafe()) {
+        // We create a new context if the previous request completes (done flat is true) or the
```
Comment (yupeng9) [inline core/server/worker/src/main/java/alluxio/worker/netty/AbstractWriteHandler.java]: done flag?

Comment (yupeng9) [reply core/server/worker/src/main/java/alluxio/worker/netty/AbstractWriteHandler.java]: I wonder why not set `mContext` to null after the request completes (i.e. when `mContext.isDoneUnsafe` is true)? Then you can simplify this logic much

File: core/server/worker/src/test/java/alluxio/worker/netty/WriteHandlerTest.java [hunk 4]
```diff
@@ -113,6 +113,36 @@ public void writeInvalidOffsetLaterRequest() throws Exception {
...
+    checkWriteResponse(PStatus.OK, writeResponse);
+    // Send second request
+    mChannel.writeInbound(newWriteRequest(0, newDataBuffer(PACKET_SIZE + 1)));
+    mChannel.writeInbound(newEofRequest(PACKET_SIZE + 1));
+    writeResponse = waitForResponse(mChannel);
+    checkWriteResponse(PStatus.OK, writeResponse);
+  }
+
+  @Test
+  public void writeCancelAndRequests() throws Exception {
```
Comment (yupeng9) [inline core/server/worker/src/test/java/alluxio/worker/netty/WriteHandlerTest.java]: Shall we have another test for `abort`?

File: core/server/worker/src/main/java/alluxio/worker/netty/AbstractWriteHandler.java (hunk 2 above)
Comment (apc999) [reply core/server/worker/src/main/java/alluxio/worker/netty/AbstractWriteHandler.java]: done

File: core/server/worker/src/main/java/alluxio/worker/netty/AbstractWriteHandler.java (hunk 3 above)
Comment (apc999) [reply core/server/worker/src/main/java/alluxio/worker/netty/AbstractWriteHandler.java]: what you are proposing is exactly the previous logic which introduces the race fixed in this PR. Consider this really rare case:

Server completes request A -> Server send OK response for request A -> Client receives OK response for request A -> Client sends new request B ->  Server receives request B  -> Server resets the context for request A.

File: core/server/worker/src/test/java/alluxio/worker/netty/WriteHandlerTest.java (hunk 4 above)
Comment (apc999) [reply core/server/worker/src/test/java/alluxio/worker/netty/WriteHandlerTest.java]: abort is not a server side control message. it is initiated from server side.
From the server side, only two types of messages: write or cancel

**Comment by apc999** [comment]
@yupeng9  your comments are either addressed or explained

**Pull request merged**